src/librustc/ty/context.rs

```diff
@@ -1756,6 +1740,16 @@ impl<'a, 'tcx> Lift<'tcx> for &'a Const<'a> {
     }
 }
 
+impl<'a, 'tcx> Lift<'tcx> for &'a mir::interpret::Allocation {
+    type Lifted = &'tcx mir::interpret::Allocation;
+    fn lift_to_tcx<'b, 'gcx>(&self, tcx: TyCtxt<'b, 'gcx, 'tcx>) -> Option<Self::Lifted> {
+        if tcx.interners.arena.in_arena(*self as *const _) {
+            return Some(unsafe { mem::transmute(*self) });
+        }
+        Some(tcx.intern_const_alloc(mir::interpret::Allocation::clone(self)))
```

Comment:
This seems different from every other `Lift` impl. I have no idea what this trait even does. Can you explain why this `impl` is special?

Reply:
`Lift` moves things from the local arena (e.g. in an infer context during type inference) to the global arena. These local arenas exist, because otherwise we start filling up the global arena with short-lived objects. Similar to how in const eval we don't allocate everything in the tcx.

This impl is "special" because it is less special. `Allocation`s have no content with any lifetimes, they are just interned in the arena. So lifting them involves cloning them and reinterning in the global arena

Reply:
Okay, makes sense, thanks!